ehr/test/src/org/labkey/test/tests/ehr/AbstractGenericEHRTest.java

```diff
@@ -300,6 +306,160 @@ private void testUserAgainstAllStates(@LoggedParam EHRUser user)
+                if (!url.getHost().equalsIgnoreCase(getURL().getHost()))
+                {
+                    log(href + " is an external link. Skipping validation.");
+                    return validUrl;
+                }
+            }
+            catch (MalformedURLException e)
+            {
+                // not a full URL so not external. Carry on.
+            }
+
+            // scope this to admin, ehr folder and subfolders
+            if (!decodedHref.contains(getContainerPath()) && !decodedHref.startsWith("/admin"))
+            {
+                log(href + " is in a different folder than the EHR folder, " + getContainerPath() + ". Skipping validation.");
+                return validUrl;
+            }
+        }
+
+        if (anchor.isDisplayed() && anchor.isEnabled())
+        {
+            clickable = true;
+
+            try
+            {
+                openLinkInNewWindowOrThrow(anchor);
+            }
+            catch (WebDriverException | IllegalStateException e)
+            {
+                clickable = false;
+            }
+
+            if (clickable)
+            {
+                // Give page time to load
+                boolean loaded = waitFor(() -> (getDriver().getCurrentUrl() != null && !getDriver().getCurrentUrl().equalsIgnoreCase("about:blank")), WAIT_FOR_PAGE);
+                assertTrue("Link " + href + " did not load in " + WAIT_FOR_PAGE + "ms.", loaded);
+
+                // Assert page is not empty and does not have errors
+                URL url = getURL();
+                assertFalse("URL " + url + " is empty.", isPageEmpty());
+                assertNoLabKeyErrors();
+
+                // assertNoLabKeyErrors does not catch all types of errors
+                assertElementNotPresent("LabKey error found for URL " + url, Locators.labkeyErrorHeading);
+
+                // record link as valid and cleanup
+                validUrl = url.toString();
+                switchToWindow(0);
+                quietlyCloseExtraWindows();
+            }
+        }
+        return validUrl;
+    }
+
+    private void validatePageLinks(Set<String> crawledLinks)
+    {
+        log("Validating links on " + getURL());
+
+        // Find all anchors in the body content area, excluding buttons and those in data regions
+        List<WebElement> anchors = getDriver().findElements(By.xpath("//div[contains(concat(' ', normalize-space(@class), ' '), ' lk-body-ct ')]//a[not(ancestor::form[@data-region-form]) and not(@role='button') and not(contains(@class, 'labkey-button'))]"));
+
+        log(anchors.size() + " possible links found.");
+        int validatedCount = 0;
+        Set<String> validLinksOnPage = new HashSet<>();
+        for (WebElement anchor : anchors)
+        {
+            // Only validate links once
+            String href = anchor.getDomAttribute("href");
+            if (href != null && validLinksOnPage.contains(href))
+                continue;
```

Comment:
Skip if link has already been crawled?
```suggestion
            if (href != null && validLinksOnPage.contains(href) || crawledLinks.contains(href))
                continue;
```

Reply:
good catch thanks